etl/src/v2/replication/apply.rs

```diff
@@ -346,101 +350,107 @@ where
-        //
-        // Early breaking can happen for example when a table sync worker has caught up with the apply worker
-        // but its batch contained more elements after the caught up element, in that case we don't
-        // want to process those elements, otherwise if we do, the apply worker will process them too
-        // causing duplicate data.
-        match state.early_break {
-            Some(BatchEarlyBreak::Break) => {
-                debug!("early break requested, stopping apply loop");
-                stop_apply_loop = true;
-
-                break;
-            }
-            Some(BatchEarlyBreak::BreakAndDiscard) => {
-                debug!(
-                    "early break and discard requested, resetting state and stopping apply loop"
-                );
-                *state = previous_state;
-                stop_apply_loop = true;
-
-                break;
-            }
-            None => {}
+        let mut end_loop = false;
+        if let Some(table_id) = skip_table {
+            end_loop |= !hook.skip_table(table_id).await?;
         }
-    }
-
-    if events_batch.is_empty() {
-        return Ok(stop_apply_loop);
-    }
 
-    // We apply the batch of events to the destination.
-    destination.write_events(events_batch).await?;
+        // At this point, the `last_commit_end_lsn` will contain the LSN of the next byte in the WAL after
+        // the last `Commit` message that was processed in this batch or in the previous ones.
+        if let Some(last_commit_end_lsn) = state.last_commit_end_lsn.take() {
+            // We also prepare the next status update for Postgres, where we will confirm that we flushed
+            // data up to this LSN to allow for WAL pruning on the database side.
+            //
+            // Note that we do this ONLY once a batch is fully saved, since that is the only place where
+            // we are guaranteed that data has been safely persisted. In all the other cases, we just update
+            // the `write_lsn` which is used by Postgres to get an acknowledgement of how far we have processed
+            // messages but not flushed them.
+            // TODO: check if we want to send `apply_lsn` as a different value.
+            state
+                .next_status_update
+                .update_flush_lsn(last_commit_end_lsn);
+            state
+                .next_status_update
+                .update_apply_lsn(last_commit_end_lsn);
+
+            // We call `process_syncing_tables` with `update_state` set to true here *after* we've received
+            // and ack for the batch from the destination. This is important to keep a consistent state.
+            // Without this order it could happen that the table's state was updated but sending the batch
+            // to the destination failed.
+            end_loop |= !hook
+                .process_syncing_tables(state.next_status_update.flush_lsn, true)
+                .await?;
```

Comment:
With the process syncing here, the system becomes much harder to reason about, but I am failing to see any alternative that would work well with batching.

Reply:
Yes, it has become harder to see the logic, but as you said I can't think of a real alternative either.